File: pine-java/src/main/java/page/liam/pine/operators/TransformCopy.java
package page.liam.pine.operators;

import page.liam.pine.AbstractOperator;
import page.liam.pine.CancellationToken;
import page.liam.pine.OperatorInput;
import page.liam.pine.OperatorOutput;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Operator: transform_copy
 * Metadata contract
 *   direction="common_to_item":
 *     CommonInput:  [<source_fields...>]
 *     ItemOutput:   [<target_fields...>]
 *   direction="item_to_item":
 *     ItemInput:    [<source_fields...>]
 *     ItemOutput:   [<target_fields...>]
 *   direction="common_to_common":
 *     CommonInput:  [<source_fields...>]
 *     CommonOutput: [<target_fields...>]
 *   direction="item_to_common":
 *     ItemInput:    [<source_field>]
 *     CommonOutput: [<target_field>]
 */
public class TransformCopy extends AbstractOperator implements page.liam.pine.ConcurrentSafe {
    private String direction;

    @Override
    public void init(Map<String, Object> params) throws Exception {
        direction = (String) params.get("direction");
        switch (direction) {
            case "common_to_item":
            case "item_to_common":
            case "common_to_common":
            case "item_to_item":
                break;
            default:
                throw new IllegalArgumentException("transform_copy: unsupported direction \"" + direction + "\"");
        }
    }

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) {
        switch (direction) {
            case "common_to_common":
                for (int i = 0; i < commonInput().size(); i++) {
                    output.setCommon(commonOutput().get(i), input.common(commonInput().get(i)));
                }
                break;

            case "common_to_item":
                for (int i = 0; i < commonInput().size(); i++) {
                    Object val = input.common(commonInput().get(i));
                    String dst = itemOutput().get(i);
                    for (int j = 0; j < input.itemCount(); j++) {
                        output.setItem(j, dst, val);
                    }
                }
                break;

            case "item_to_item":
                for (int i = 0; i < itemInput().size(); i++) {
                    String src = itemInput().get(i);
                    String dst = itemOutput().get(i);
                    for (int j = 0; j < input.itemCount(); j++) {
                        output.setItem(j, dst, input.item(j, src));
                    }
                }
                break;

            case "item_to_common":
                for (int i = 0; i < itemInput().size(); i++) {
                    String src = itemInput().get(i);
                    List<Object> vals = new ArrayList<>();
                    for (int j = 0; j < input.itemCount(); j++) {
                        vals.add(input.item(j, src));
                    }
                    output.setCommon(commonOutput().get(i), vals);
                }
                break;
        }
    }
}

File: pine-python/pine/operators/transform_copy.py
from __future__ import annotations

from pine.cancellation import CancellationToken
from pine.operator import (
    AbstractOperator,
    ConcurrentSafe,
    ConsumesRowSet,
    OperatorInput,
    OperatorOutput,
    OperatorParams,
)


class TransformCopy(AbstractOperator, ConcurrentSafe, ConsumesRowSet):
    def __init__(self):
        self._direction = ""

    def init(self, params: OperatorParams):
        self._direction = params.get_string("direction")
        if self._direction not in (
            "common_to_item", "item_to_common", "common_to_common", "item_to_item"
        ):
            raise ValueError(
                f'transform_copy: unsupported direction "{self._direction}"'
            )

    def execute(
        self, token: CancellationToken, input_: OperatorInput,
        output: OperatorOutput,
    ) -> None:
        if self._direction == "common_to_common":
            for i in range(len(self.common_input())):
                output.set_common(
                    self.common_output()[i],
                    input_.common(self.common_input()[i]),
                )

        elif self._direction == "common_to_item":
            for i in range(len(self.common_input())):
                val = input_.common(self.common_input()[i])
                dst = self.item_output()[i]
                for j in range(input_.item_count()):
                    output.set_item(j, dst, val)

        elif self._direction == "item_to_item":
            for i in range(len(self.item_input())):
                src = self.item_input()[i]
                dst = self.item_output()[i]
                for j in range(input_.item_count()):
                    output.set_item(j, dst, input_.item(j, src))

        elif self._direction == "item_to_common":
            for i in range(len(self.item_input())):
                src = self.item_input()[i]
                vals: list = []
                for j in range(input_.item_count()):
                    vals.append(input_.item(j, src))
                output.set_common(self.common_output()[i], vals)
